File: ts/routes/(gre)/session-completion.ts
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

import type { DashboardTopicInsight } from "@generated/anki/brainlift_pb";

import {
    GRE_CTA_PRACTICE,
    GRE_CTA_REVIEW,
    GRE_CTA_STUDY_PLAN,
    type GreNavAction,
    greNavAction,
    greNavItem,
} from "./gre-navigation";
import {
    extraStudyActionLabel,
    extraStudyDetail,
    nextReviewScheduleLabel,
} from "./review/extra-study";
import { formatPercent } from "./score-format";

export type SessionAttemptRecord = {
    topic: string;
    correct: boolean;
};

export type SessionCompletionRow = {
    label: string;
    value: string;
};

export type SessionCompletionSummary = {
    headline: string;
    subline: string;
    rows: SessionCompletionRow[];
    nextAction?: GreNavAction;
    nextActionDetail?: string;
    secondaryAction?: GreNavAction;
};

type TopicStats = Map<string, { correct: number; total: number }>;

function topicStats(attempts: SessionAttemptRecord[]): TopicStats {
    const stats: TopicStats = new Map();
    for (const attempt of attempts) {
        const topic = attempt.topic.trim() || "Unknown topic";
        const current = stats.get(topic) ?? { correct: 0, total: 0 };
        stats.set(topic, {
            correct: current.correct + (attempt.correct ? 1 : 0),
            total: current.total + 1,
        });
    }
    return stats;
}

function strongestFromStats(stats: TopicStats): string | null {
    let best: { topic: string; accuracy: number; correct: number } | null = null;
    for (const [topic, value] of stats) {
        const accuracy = value.correct / value.total;
        if (
            !best
            || accuracy > best.accuracy
            || (accuracy === best.accuracy && value.correct > best.correct)
        ) {
            best = { topic, accuracy, correct: value.correct };
        }
    }
    return best?.topic ?? null;
}

function weakestFromStats(stats: TopicStats, strongest: string | null): string | null {
    if (stats.size === 0) {
        return null;
    }
    let worst: { topic: string; accuracy: number; total: number } | null = null;
    for (const [topic, value] of stats) {
        const accuracy = value.correct / value.total;
        if (
            !worst
            || accuracy < worst.accuracy
            || (accuracy === worst.accuracy && value.total > worst.total)
        ) {
            worst = { topic, accuracy, total: value.total };
        }
    }
    if (!worst) {
        return null;
    }
    if (strongest && worst.topic === strongest && stats.size === 1) {
        return worst.accuracy < 1 ? worst.topic : null;
    }
    if (strongest && worst.topic === strongest) {
        return null;
    }
    return worst.topic;
}

function topicScore(topic: DashboardTopicInsight): number {
    if (topic.practiceAccuracy !== undefined) {
        return topic.practiceAccuracy;
    }
    if (topic.memoryScore !== undefined) {
        return topic.memoryScore;
    }
    return topic.covered ? 0.5 : 0;
}

function strongestDashboardTopic(
    recommendedTopics: DashboardTopicInsight[],
    weakTopics: DashboardTopicInsight[],
): string | null {
    const candidates = topicsWithStudyEvidence([...recommendedTopics, ...weakTopics]);
    if (candidates.length === 0) {
        return null;
    }
    const sorted = [...candidates].sort((a, b) => topicScore(b) - topicScore(a));
    return sorted[0]?.displayName ?? null;
}

function topicsWithStudyEvidence(
    topics: DashboardTopicInsight[],
): DashboardTopicInsight[] {
    return topics.filter(
        (topic) =>
            topic.studiedCards > 0
            || topic.memoryScore !== undefined
            || topic.practiceAccuracy !== undefined,
    );
}

function lacksGreFlashcardEvidence(studiedCards: number, coveredLeafCount: number): boolean {
    return studiedCards === 0 && coveredLeafCount === 0;
}

export type PracticeSessionSummaryOptions = {
    focusTopicName?: string;
    focusComplete?: boolean;
    flashcardScheduleHint?: string;
};

export function buildPracticeSessionSummary(
    attempts: SessionAttemptRecord[],
    options?: PracticeSessionSummaryOptions,
): SessionCompletionSummary {
    const total = attempts.length;
    const correct = attempts.filter((attempt) => attempt.correct).length;
    const stats = topicStats(attempts);
    const strongest = strongestFromStats(stats);
    const weakest = weakestFromStats(stats, strongest);

    const rows: SessionCompletionRow[] = [
        {
            label: "Questions answered",
            value: String(total),
        },
    ];

    if (total > 0) {
        rows.push({
            label: "Accuracy",
            value: formatPercent((correct / total) * 100),
        });
    }
    if (strongest) {
        rows.push({ label: "Strongest topic", value: strongest });
    }
    if (weakest) {
        rows.push({ label: "Focus next", value: weakest });
    }

    if (options?.focusComplete) {
        const topic = options.focusTopicName ?? "this topic";
        if (options.flashcardScheduleHint) {
            rows.push({ label: "Flashcard review", value: options.flashcardScheduleHint });
        }
        const dashboardAction = greNavAction(greNavItem("dashboard"));
        dashboardAction.label = GRE_CTA_STUDY_PLAN;
        const practiceAction = greNavAction(greNavItem("practice"));
        practiceAction.label = "Practice again";
        return {
            headline: "Focus complete",
            subline: `You finished today's mission for ${topic}.`,
            rows,
            nextAction: dashboardAction,
            nextActionDetail: options.flashcardScheduleHint
                ? `${options.flashcardScheduleHint}. Head back to your dashboard for what's next.`
                : "Head back to your dashboard for what's next.",
            secondaryAction: practiceAction,
        };
    }

    let nextAction = greNavAction(greNavItem("practice"));
    nextAction.label = "Practice again";
    let nextActionDetail = "Run another short set while this material is fresh.";

    if (weakest) {
        nextActionDetail = `Your weakest area this session was ${weakest}. Another short set there will help it stick.`;
    } else if (total > 0 && correct === total) {
        nextAction = greNavAction(greNavItem("study"));
        nextAction.label = GRE_CTA_REVIEW;
        nextActionDetail = "Strong accuracy — reinforce recall with a quick flashcard review.";
    }

    return {
        headline: "Session complete",
        subline: "Here's how this set went.",
        rows,
        nextAction,
        nextActionDetail,
        secondaryAction: greNavAction(greNavItem("dashboard")),
    };
}

export function buildStudyCaughtUpSummary(input: {
    weakTopics: DashboardTopicInsight[];
    recommendedTopics: DashboardTopicInsight[];
    dueTotal: number;
    deckName: string;
    studiedCards: number;
    coveredLeafCount: number;
    extraStudyAvailable?: number;
    availableNewCount?: number;
    nextReviewInDays?: number;
}): SessionCompletionSummary {
    if (lacksGreFlashcardEvidence(input.studiedCards, input.coveredLeafCount)) {
        const startStudy = greNavAction(greNavItem("study"));
        startStudy.label = GRE_CTA_REVIEW;
        const practiceAction = greNavAction(greNavItem("practice"));
        practiceAction.label = GRE_CTA_PRACTICE;
        return {
            headline: "Your GRE flashcards are ready",
            subline: "GRE Atlas includes built-in flashcards — no Anki import needed.",
            rows: [
                {
                    label: "Flashcards reviewed",
                    value: "0",
                },
            ],
            nextAction: startStudy,
            nextActionDetail: "Tap Start review to open your first cards and begin building memory evidence.",
            secondaryAction: practiceAction,
        };
    }

    const studiedWeakTopics = topicsWithStudyEvidence(input.weakTopics);
    const weakest = studiedWeakTopics[0]?.displayName ?? null;
    const strongest = strongestDashboardTopic(
        input.recommendedTopics,
        input.weakTopics,
    );

    const rows: SessionCompletionRow[] = [
        {
            label: "Cards due now",
            value: String(input.dueTotal),
        },
        {
            label: "Flashcards reviewed",
            value: String(input.studiedCards),
        },
    ];
    if (strongest) {
        rows.push({ label: "Strongest area", value: strongest });
    }
    if (weakest) {
        rows.push({ label: "Focus next", value: weakest });
    } else if (input.studiedCards === 0) {
        rows.push({
            label: "Flashcard history",
            value: "Not enough yet",
        });
    }

    const scheduleLabel = nextReviewScheduleLabel(input.nextReviewInDays);
    if (input.dueTotal === 0 && scheduleLabel && (input.extraStudyAvailable ?? 0) === 0) {
        rows.push({ label: "Next flashcard review", value: scheduleLabel });
    }

    const extraAvailable = input.extraStudyAvailable ?? 0;
    if (extraAvailable > 0) {
        const studyAhead: GreNavAction = {
            label: extraStudyActionLabel(extraAvailable),
            bridge: "greStartExtraReview",
            href: "/review",
        };
        const practiceAction = greNavAction(greNavItem("practice"));
        practiceAction.label = GRE_CTA_PRACTICE;
        return {
            headline: input.dueTotal === 0 ? "Review complete" : "Session complete",
            subline: input.dueTotal === 0
                ? "You're caught up on flashcards due right now."
                : "Nice pause point — you can pick up remaining cards later.",
            rows,
            nextAction: studyAhead,
            nextActionDetail: extraStudyDetail(extraAvailable),
            secondaryAction: practiceAction,
        };
    }

    return {
        headline: input.dueTotal === 0 ? "Review complete" : "Session complete",
        subline: input.dueTotal === 0
            ? "You're caught up on flashcards due right now."
            : "Nice pause point — you can pick up remaining cards later.",
        rows,
    };
}

File: mobile/ios/GREAtlasCompanion/Engine/SessionCompletion.swift
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

import Foundation

struct PracticeAttemptRecord: Equatable {
    let topic: String
    let correct: Bool
}

struct SessionCompletionRow: Equatable {
    let label: String
    let value: String
}

struct SessionCompletionSummary: Equatable {
    let headline: String
    let subline: String
    let rows: [SessionCompletionRow]
    let nextActionLabel: String
    let nextActionDetail: String
    let nextActionTab: GreNavTab?
    let nextActionTopicId: String?
    let usesExtraStudy: Bool
    let secondaryActionLabel: String
    let secondaryActionTab: GreNavTab
}

enum SessionCompletionBuilder {
    static func practiceSummary(from attempts: [PracticeAttemptRecord]) -> SessionCompletionSummary {
        let total = attempts.count
        let correct = attempts.filter(\.correct).count
        let stats = topicStats(from: attempts)
        let strongest = strongestTopic(from: stats)
        let weakest = weakestTopic(from: stats, strongest: strongest)

        var rows: [SessionCompletionRow] = [
            SessionCompletionRow(label: "Questions answered", value: "\(total)"),
        ]
        if total > 0 {
            rows.append(
                SessionCompletionRow(
                    label: "Accuracy",
                    value: ScoreFormat.formatPercent(Double(correct) / Double(total) * 100)
                )
            )
        }
        if let strongest {
            rows.append(SessionCompletionRow(label: "Strongest topic", value: strongest))
        }
        if let weakest {
            rows.append(SessionCompletionRow(label: "Focus next", value: weakest))
        }

        var nextActionLabel = "Practice again"
        var nextActionTab: GreNavTab? = .practice
        let nextActionTopicId: String? = nil
        var nextActionDetail = "Run another short set while this material is fresh."

        if let weakest {
            nextActionDetail =
                "Your weakest area this session was \(weakest). Another short set there will help it stick."
        } else if total > 0, correct == total {
            nextActionLabel = "Review flashcards"
            nextActionTab = .study
            nextActionDetail = "Strong accuracy — reinforce recall with a quick flashcard review."
        }

        return SessionCompletionSummary(
            headline: "Session complete",
            subline: "Here's how this set went.",
            rows: rows,
            nextActionLabel: nextActionLabel,
            nextActionDetail: nextActionDetail,
            nextActionTab: nextActionTab,
            nextActionTopicId: nextActionTopicId,
            usesExtraStudy: false,
            secondaryActionLabel: nextActionLabel == "Practice again" ? "View study plan" : "Practice again",
            secondaryActionTab: nextActionLabel == "Practice again" ? .dashboard : .practice
        )
    }

    static func studyCaughtUpSummary(
        weakTopic: GreTopicInsightView?,
        recommendedTopics: [GreTopicInsightView],
        dueTotal: UInt,
        studiedCards: UInt,
        extraStudyAvailable: UInt = 0,
        nextReviewInDays: UInt? = nil
    ) -> SessionCompletionSummary {
        let weakest = weakTopic?.displayName
        let strongest = strongestDashboardTopic(recommendedTopics: recommendedTopics, weakTopic: weakTopic)

        var rows: [SessionCompletionRow] = [
            SessionCompletionRow(label: "Cards due now", value: "\(dueTotal)"),
            SessionCompletionRow(label: "Flashcards reviewed", value: "\(studiedCards)"),
        ]
        if let strongest {
            rows.append(SessionCompletionRow(label: "Strongest area", value: strongest))
        }
        if let weakest {
            rows.append(SessionCompletionRow(label: "Focus next", value: weakest))
        }
        if dueTotal == 0, extraStudyAvailable == 0, let nextReviewInDays {
            rows.append(
                SessionCompletionRow(
                    label: "Next flashcard review",
                    value: nextReviewScheduleLabel(days: nextReviewInDays)
                )
            )
        }

        if extraStudyAvailable > 0 {
            let label = extraStudyAvailable == 1
                ? "Study 1 more card"
                : "Study \(extraStudyAvailable) more cards"
            return SessionCompletionSummary(
                headline: dueTotal == 0 ? "Review complete" : "Session complete",
                subline: dueTotal == 0
                    ? "You're caught up on flashcards due right now."
                    : "Nice pause point — you can pick up remaining cards later.",
                rows: rows,
                nextActionLabel: label,
                nextActionDetail:
                    "Unlock up to \(extraStudyAvailable) new flashcard\(extraStudyAvailable == 1 ? "" : "s") today (20 max per day) to build memory evidence without cramming.",
                nextActionTab: .study,
                nextActionTopicId: nil,
                usesExtraStudy: true,
                secondaryActionLabel: "Practice questions",
                secondaryActionTab: .practice
            )
        }

        var nextActionLabel = "Practice questions"
        var nextActionTab: GreNavTab? = .practice
        var nextActionTopicId: String? = nil
        var nextActionDetail = "Keep momentum with a short practice set."

        if let weakTopic {
            nextActionLabel = "Practice weak area"
            nextActionTopicId = weakTopic.topicId
            nextActionDetail = "Your next best step is a few questions on \(weakTopic.displayName)."
        } else {
            nextActionLabel = "View study plan"
            nextActionTab = nil
            nextActionDetail = "Open your study plan to pick the next focus area."
        }

        return SessionCompletionSummary(
            headline: dueTotal == 0 ? "Review complete" : "Session complete",
            subline: dueTotal == 0
                ? "You're caught up on flashcards due right now."
                : "Nice pause point — you can pick up remaining cards later.",
            rows: rows,
            nextActionLabel: nextActionLabel,
            nextActionDetail: nextActionDetail,
            nextActionTab: nextActionTab,
            nextActionTopicId: nextActionTopicId,
            usesExtraStudy: false,
            secondaryActionLabel: "View study plan",
            secondaryActionTab: .dashboard
        )
    }

    private static func nextReviewScheduleLabel(days: UInt) -> String {
        switch days {
        case 0:
            return "Flashcards due today"
        case 1:
            return "Flashcards due tomorrow"
        default:
            return "Flashcards due in \(days) days"
        }
    }

    private static func topicStats(from attempts: [PracticeAttemptRecord]) -> [String: (correct: Int, total: Int)] {
        var stats: [String: (correct: Int, total: Int)] = [:]
        for attempt in attempts {
            let topic = attempt.topic.trimmingCharacters(in: .whitespacesAndNewlines)
            let key = topic.isEmpty ? "Unknown topic" : topic
            var current = stats[key] ?? (0, 0)
            current.total += 1
            if attempt.correct { current.correct += 1 }
            stats[key] = current
        }
        return stats
    }

    private static func strongestTopic(from stats: [String: (correct: Int, total: Int)]) -> String? {
        stats.max { lhs, rhs in
            let leftAccuracy = Double(lhs.value.correct) / Double(lhs.value.total)
            let rightAccuracy = Double(rhs.value.correct) / Double(rhs.value.total)
            if leftAccuracy == rightAccuracy {
                return lhs.value.correct < rhs.value.correct
            }
            return leftAccuracy < rightAccuracy
        }?.key
    }

    private static func weakestTopic(
        from stats: [String: (correct: Int, total: Int)],
        strongest: String?
    ) -> String? {
        guard !stats.isEmpty else { return nil }
        guard let worst = stats.min(by: { lhs, rhs in
            let leftAccuracy = Double(lhs.value.correct) / Double(lhs.value.total)
            let rightAccuracy = Double(rhs.value.correct) / Double(rhs.value.total)
            if leftAccuracy == rightAccuracy {
                return lhs.value.total < rhs.value.total
            }
            return leftAccuracy < rightAccuracy
        }) else { return nil }

        if stats.count == 1 {
            let accuracy = Double(worst.value.correct) / Double(worst.value.total)
            return accuracy < 1 ? worst.key : nil
        }
        if worst.key == strongest { return nil }
        return worst.key
    }

    private static func topicScore(_ topic: GreTopicInsightView) -> Double {
        if let accuracy = topic.practiceAccuracy { return accuracy }
        if let memory = topic.memoryScore { return memory }
        return topic.studyLabel.isEmpty ? 0 : 0.5
    }

    private static func strongestDashboardTopic(
        recommendedTopics: [GreTopicInsightView],
        weakTopic: GreTopicInsightView?
    ) -> String? {
        var candidates = recommendedTopics
        if let weakTopic { candidates.append(weakTopic) }
        return candidates.max(by: { topicScore($0) < topicScore($1) })?.displayName
    }
}
